lib/src/async_io.rs

```diff
@@ -7,28 +7,28 @@ use {
     futures::{FutureExt, TryFutureExt},
     std::time::Duration,
     tokio::time::timeout,
-    wiggle::GuestPtr,
+    wiggle::{GuestMemory, GuestPtr},
 };
 
 #[wiggle::async_trait]
 impl FastlyAsyncIo for Session {
-    async fn select<'a>(
+    async fn select(
         &mut self,
-        handles: &GuestPtr<'a, [AsyncItemHandle]>,
+        memory: &mut GuestMemory<'_>,
+        handles: GuestPtr<[AsyncItemHandle]>,
         timeout_ms: u32,
     ) -> Result<u32, Error> {
-        let handles = GuestPtr::<'a, [u32]>::new(handles.mem(), handles.offset())
-            .as_slice()?
-            .ok_or(Error::SharedMemory)?;
+        let handles = handles.cast::<[u32]>();
         if handles.len() == 0 && timeout_ms == 0 {
             return Err(Error::InvalidArgument);
         }
 
         let select_fut = self
             .select_impl(
-                handles
-                    .iter()
-                    .copied()
+                memory
+                    // TODO: wiggle only supports slices of u8 in 22.0.0
```

Comment:
This comment is confusing me a bit. Based on the comment, I would guess that `memory.to_vec(handles)` returns a `Vec<u8>`, which would be a problem because that would mean we'd be creating `AsyncItemHandle`s from bytes instead of 32-bit values. But if I look at the `wiggle` docs, it looks like `memory_to_vec(handles)` is going to keep the type `T` (in this case, `u32`), so we'll be doing the right thing here.

So ... what's up?

Reply:
Yep, this needs to be reworded. The issue is that the `as_slice` function on `GuestMemory` will only handle `GuestPtr<[u8]>`, and not `GuestPtr<[T]>`. This was just an oversight, and in the future we can avoid copying with `to_vec` here.

Reply:
(Just to be really clear: this does the right thing when producing a vector of `u32`, but it copies in the process)